src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs

```diff
@@ -773,29 +763,127 @@ private void TasView_MouseDown(object sender, MouseEventArgs e)
+		/// </summary>
+		private void BeginBatchEdit()
+		{
+			_batchEditing = true;
+		}
+		/// <returns>Returns true if the input list was redrawn.</returns>
+		private bool EndBatchEdit()
+		{
+			_batchEditing = false;
+			if (_batchEditMinFrame != -1)
+			{
+				return FrameEdited(_batchEditMinFrame);
+			}
+
+			return false;
+		}
+
+		public void BatchEdit(Action action)
+		{
+			// This is only caled from Lua.
+			_editIsFromLua = true;
+			BeginBatchEdit();
+			action();
+			EndBatchEdit();
+			_editIsFromLua = false;
+		}
+
+		/// <summary>
+		/// Disables recording mode, ensures we are in the greenzone, and does autorestore if needed.
+		/// If a mouse button is down, only tracks the edit so we can do this stuff on mouse up.
+		/// </summary>
+		/// <param name="frame">The frame that was just edited, or the earliest one if multiple were edited.</param>
+		/// <returns>Returns true if the input list was redrawn.</returns>
+		public bool FrameEdited(int frame)
+		{
+			GreenzoneInvalidatedCallback?.Invoke(frame); // lua callback
+
+			if (CurrentTasMovie.LastEditWasRecording)
+				return false;
+
+			bool needsRefresh = !_batchEditing;
+			if (MouseButtonHeld || _batchEditing)
+			{
+				if (_batchEditMinFrame == -1)
+				{
+					_batchEditMinFrame = frame;
+				}
+				else
+				{
+					_batchEditMinFrame = Math.Min(_batchEditMinFrame, frame);
+				}
+			}
+			else
+			{
+				if (!_editIsFromLua)
+				{
+					// Lua users will want to preserve recording mode.
+					TastudioPlayMode(true);
+				}
+
+				if (Emulator.Frame > frame)
+				{
+					if (_shouldMoveGreenArrow)
+					{
+						RestorePositionFrame = _seekingTo != -1 ? _seekingTo : Emulator.Frame;
+						// Green arrow should not move again until the user changes frame.
+						// This means any state load or unpause/frame advance/seek, that is not caused by an input edit.
+						// This is so that the user can make multiple edits with auto restore off, in any order, before a manual restore.
+						_shouldMoveGreenArrow = false;
+					}
+
+					_seekingByEdit = true; // must be before GoToFrame (it will load a state, and state loading checks _seekingByEdit)
+					GoToFrame(frame);
+					if (Settings.AutoRestoreLastPosition)
+					{
+						RestorePosition();
+					}
+					_seekingByEdit = true; // must be after GoToFrame & RestorePosition too (they'll set _seekingByEdit to false)
+
+					needsRefresh = false; // Refresh will happen via GoToFrame.
+				}
+				_batchEditMinFrame = -1;
+			}
+
+			if (needsRefresh)
+			{
+				if (TasView.IsPartiallyVisible(frame) || frame < TasView.FirstVisibleRow)
+				{
+					// frame < FirstVisibleRow: Greenzone in visible rows has been invalidated
+					RefreshDialog();
+					return true;
+				}
+				else if (TasView.RowCount != CurrentTasMovie.FrameCount + 1)
+				{
+					// Row count must always be kept up to date even if last row is not directly visible.
+					TasView.RowCount = CurrentTasMovie.InputLogLength + 1;

```

Comment:
Is this meant to be a different prop to the one in the condition?

Reply:
No, I didn't even realize there were two such properties. What is the theoretical difference? It's not explained in the property comments and their implementations are identical.